library/std/src/error.rs

```diff
@@ -488,6 +489,23 @@ impl<T: Error> Error for Box<T> {
     }
 }
 
+#[stable(feature = "arc_error", since = "1.51.0")]
+impl<T: Error> Error for Arc<T> {
```

Comment:
```suggestion
impl<T: Error + ?Sized> Error for Arc<T> {
```

[unlike Box](https://github.com/rust-lang/rust/pull/75180#issuecomment-670292054) we should be able to add the `?Sized` bound.

Reply:
That would prevent a `From` impl for `E: Error` in the future, we okay with that?

Reply:
I'm not sure there would be many `?Sized` errors in practice. Could this bound be relaxed (with `?Sized`) at a later date and still be backwards-compatible?

Reply:
I mean, it might be that it does no harm if we assume a future specialization feature would allow `From` impl for `E: Error` in the future.

Reply:
@derekchiang we couldn't put a `?Sized` on in the future backwards compatibly. This is a problem with `Box` right now. This causes overlap for the impls `From<T> for T` and `From<E: Error> for Arc<dyn Error + ...>`. It would become ambiguous when converting `Arc<dyn Error>` into an `Arc<dyn Error>` whether or not you should do a unit conversion or wrap the Arc in another Arc.

As for specialization, I'm not as sure but as far as I know it won't help here. I've looked this up in the past and from previous research my understanding is that this would require rustc to implement [the lattice rule](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md#the-lattice-rule), but afaik there are no plans to implement support for lattice specialization. That said, I've never had this confirmed by @rust-lang/wg-traits, so this might be on the roadmap even if it's not in the RFC, or I could be misunderstanding the RFC.

Reply:
This stuff is extremely nuanced, thankyou for taking the time to work though it all.

Reply:
I think ideally we'd want the `?Sized` bound on there. The fact that `Box<dyn Error>`, which is a very common pattern, doesn't implement `Error` is something we regret. So I think aligning this impl to [the one for `&'a E: Error + ?Sized`](https://github.com/rust-lang/rust/pull/75180/files) is better than aligning it to `Box<E: Error>`.

Reply:
Agreed, @derekdreery, could you go ahead and apply this suggestion